_topics/en/htlc.md

```diff
@@ -0,0 +1,133 @@
+---
+title: Hash Time Locked Contract (HTLC)
+
+## Optional.  Shorter name to use for reference style links e.g., "foo"
+## will allow using the link [topic foo][].  Not case sensitive
+shortname: htlc
+
+## Optional.  An entry will be added to the topics index for each alias
+#aliases:
+#  - Foo
+
+## Required.  At least one category to which this topic belongs.  See
+## schema for options
+categories:
+  - Contract Protocols
+  - Lightning Network
+
+## Required.  Use Markdown formatting.  Only one paragraph.  No links allowed.
+## Should be less than 500 characters
+excerpt: >
+  **Hash Time Locked Contracts (HTLCs)** are conditional payments used
+  in LN payment channels, cross-chain atomic swaps, same-chain
+  coinswaps, zero-knowledge contingent payments, and other contract
+  protocols.
+
+## Optional.  Use Markdown formatting.  Multiple paragraphs.  Links allowed.
+extended_summary: |
+  HTLCs have two fundamental clauses: a payment clause secured with a
+  *hash lock* and a refund clause secured with a *time lock*.  To open a
+  hash lock and claim a payment, the receiver needs to reveal the
+  preimage to a hash digest encoded in the contract.  To open a time
+  lock and receive a refund, the spender needs to wait until after a
+  certain time encoded in the contract.
+
+  Because revealed preimages and past times don't uniquely identify the
+  person who should receive either the payment, HTLCs are only secure if
+  they also require a unique signature matching the public key of either
+  the spender (refundee) or the intended receiver.  That makes the
+  minimal HTLC look something [like this][htlc in minsc] in the
+  [Minsc][] scripting language:
+
+  ```hack
+  (pk(Receiver) && sha256(H)) || (pk(Refundee) && older(10))
+  ```
+
+  ### History
+
+  Hash locks and time locks as independent features were clearly
+  designed into the original version of Bitcoin.  As far as we're aware,
+  the earliest description combining the two to create a conditional
+  payment---what'd we now call an HTLC---is a set of posts ([1][maxwell
+  hashlock], [2][maxwell timelock]) from July 2012. However, it's
+  possible that a [December 2010 post][nakamoto risk free trade] was
+  alluding to the same basic idea when mentioning a "cryptographically
+  [...] risk free trade".
+
+  ### Future
+
+  [Point Time Lock Contracts][topic ptlc] (PTLCs) perform the same
+  function as HTLCs but can provide better privacy, use less block
+  space, and prevent routing interception.  As a downside, they depend
+  on [signature adaptors][topic adaptor signatures] which can only
+  implemented using Bitcoin's existing ECDSA signatures by making
+  additional security assumptions or by using an `OP_CHECKMULTISIG`
```

Comment:
I believe it's possible to implement signature adaptors in ECDSA _without_ additional security assumptions (e.g. by using https://eprint.iacr.org/2019/503.pdf), but I'm certainly not an expert. Perhaps rewording this to "by using complex multi-party protocols" would be more accurate.

I see that the adaptor signatures topic page also states that additional security assumptions are required.

Reply:
Interesting!  Instead of dropping "additional security assumptions", we could possibly add an additional problem like "slow".  E.g. the paper you cites shows its protocol and the protocol it's comparing to requiring several seconds for key generation on a 256-bit curve with 128 bits security:

![2020-09-25-07_41_53_959745525](https://user-images.githubusercontent.com/61096/94263261-c5192b80-ff02-11ea-93a2-4ddce4fad8bf.png)

(And that's on a 3.6 GHz x86_64 machine (single core); if multisig keys need to be generated on a hardware wallet, that seems essentially impractical with current Trezor, Ledger, or ColdCard devices.)

Reply:
> I believe it's possible to implement signature adaptors in ECDSA _without_ additional security assumptions (e.g. by using https://eprint.iacr.org/2019/503.pdf), but I'm certainly not an expert.

This paper *does* involve very bespoke assumptions about *class groups of imaginary quadratic fields* (which are among the wildest assumptions you can make). However, there are papers that do multi-party ECDSA without extra assumptions over ECDSA. See https://eprint.iacr.org/2018/499.pdf.

The question is: Does (non-multiparty) ECDSA adaptor signatures require additional assumptions over ECDSA itself. My research says yes because the scheme has a bug: each ECDSA adaptor signature leaks a Diffie-Hellman key between the signing and encryption key (referred to here as `T`). In https://eprint.iacr.org/2020/476.pdf they fix this by making the receiver always provide a proof of knowledge for `T` i.e. they could have produced the Diffie-Hellman key without the adaptor signature since they know `t` so it doesn't matter. If you do this then it requires no new assumptions.

Now in practice, this is all moot because by using single-signer adaptor ECDSA adaptor signatures you are using a script 2-of-2 and therefore p2wsh which means you are relying on collision resistance of HASH160 which arguably the worst assumption of all.

Having said all that, I think the current paragraph is good as it is. It feels like working in this information about the bug in ECDSA adaptor signatures is too much for the intended audience. so LGTM!

Reply:
Thanks Lloyd! Really appreciate the explanation.

Reply:
@llfourn

Awesome explanation, thanks!  One note:

> using single-signer adaptor ECDSA adaptor signatures you are using a script 2-of-2 and therefore p2wsh which means you are relying on collision resistance of HASH160

P2WSH uses SHA256.  See [BIP141](https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#Witness_program), "If the version byte is 0, and the witness program is 32 bytes:  [...] SHA256 of the witnessScript must match the 32-byte witness program."

Reply:
Thanks for the correction @harding! That's good news!